### PR TITLE
HTTP Helper

### LINKED ISSUE #90: Make frontend-backend communication generic
ATM the HTTP-communication is handled completely in QueryServiceImpl.ts.
We need a generic solution (-> `HttpService`) for the network communication so common details like base-url, headers are handled only once.

Example calls of that:
```ts
// returns Promise<MyReturnType> and calls GET http://example.com/api/getNodeDetails
httpService.get<MyReturnType>('getNodeDetails'); 

// return Promise<MyReturnType> and calls POST http://example.com/api/queryAll with Body myBody
httpService.post<MyReturnType>('queryAll', { body: myBody });

// sets header 'MyHeader' to 'myValue'
httpService.post<MyReturnType>('queryAll', { body: myBody, headers: {'MyHeader': 'myValue' }});
```

### PR DESCRIPTION
**Pull request by CatoLeanTruetschel**
This PR implements a generic HTTP Helper.
Closes #90

### CODE REVIEW

File: frontend/src/services/HTTPHelper.ts [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+ * @returns The built options.
+ */
+function buildOptions(options: Partial<HTTPHelperOptions>): HTTPHelperOptions {
+  let { baseUri } = options;
+
+  // As per the spec of HTTPHelperOptions, we use the base URI of the source the frontend
+  // was loaded from, if no base uri was specified in the options.
+  if (!baseUri) {
+    const { location } = window;
+    baseUri = `${location.protocol}//${location.host}/${
```
Comment (CatoLeanTruetschel) [inline frontend/src/services/HTTPHelper.ts:102]: Add a comment that `location.protocol` includes the colon at the end.

File: frontend/src/services/HTTPHelper.ts [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+    // Check the HTTP response status code to be in the success range.
+    if (
+      httpResponse.status >= 200 &&
+      httpResponse.status <= 299 &&
+      httpResponse.result
+    ) {
+      return httpResponse.result;
+    }
+
+    throw Error(httpResponse.statusText);
```
Comment (CatoLeanTruetschel) [inline frontend/src/services/HTTPHelper.ts:280]: Add a dedicated error type.

File: frontend/src/services/HTTPHelper.ts [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+type URLQueryValueType = number | string | (number | string)[];
+
+interface URLQuery {
+  [key: string]: URLQueryValueType;
+}
+
+export abstract class HTTPRequest {
+  // TODO: Why is this constructor useless? It protects the class to be consistent.
+  // eslint-disable-next-line no-useless-constructor
+  public constructor(
```
Comment (joluj) [inline frontend/src/services/HTTPHelper.ts:24]: try to add this to your `.eslintrc.json`
```jsonc
"rules": {
    "no-useless-constructor": "off", // Disable eslint no-useless-constructor
    "@typescript-eslint/no-useless-constructor": ["error"] // enable typescript no-useless-constructor
}
```
Tthe current eslint apparently does not know that you can put class attributes directly in the constructor.
Maybe you have to add @typescript-eslint to the package.json. See backend.

File: frontend/src/services/HTTPHelper.ts [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+    super(headers, query);
+  }
+}
+
+export class HTTPPOSTRequest extends HTTPRequest {
+  // TODO: Why is this constructor useless? It protects the class to be consistent.
+  // eslint-disable-next-line no-useless-constructor
+  public constructor(
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
+    public readonly body: any,
```
Comment (joluj) [inline frontend/src/services/HTTPHelper.ts:43]: use `unknown` instead of `any`

File: frontend/src/services/HTTPHelper.ts [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+ * @param options The partial options that very specified externally.
+ * @returns The built options.
+ */
+function buildOptions(options: Partial<HTTPHelperOptions>): HTTPHelperOptions {
+  let { baseUri } = options;
+
+  // As per the spec of HTTPHelperOptions, we use the base URI of the source the frontend
+  // was loaded from, if no base uri was specified in the options.
+  if (!baseUri) {
+    const { location } = window;
```
Comment (joluj) [inline frontend/src/services/HTTPHelper.ts:101]: Isn't that just `window.location.origin`?
Example:
`https://github.com/amosproj/amos-ss2021-project2-context-map/pull/101/files#submit-review` -> `https://github.com`

File: frontend/src/services/HTTPHelper.ts [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+      request.body,
+      cancellation
+    );
+  }
+
+  private executeRequest<TResult>(
+    url: string | URL,
+    request: HTTPRequest,
+    method: 'post' | 'get' | 'head' | 'put' | 'delete' | 'patch',
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
```
Comment (joluj) [inline frontend/src/services/HTTPHelper.ts:144]: use `unknown` instead of `any`

File: frontend/src/services/HTTPHelper.ts [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+      httpClient.open(method, parsedURL.href, true);
+
+      // Process the request-headers
+      // Walk over all attributes of the 'headers' object and set the key-value pairs as request header.
+      const keys = Object.keys(headers);
+      for (let i = 0; i < keys.length; i += 1) {
+        const key = keys[i];
+        const value = headers[key];
+
+        if (typeof value === 'string') {
```
Comment (joluj) [inline frontend/src/services/HTTPHelper.ts:172]: why don't we define `HTTPHeaderCollection` as
```ts
interface HTTPHeaderCollection {
  [key: string]: string; // removed  "| undefined"
}
```
Then this is tested at compile time

File: frontend/src/services/HTTPHelper.ts [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+      // This does not necessarily mean, that everything was successful.
+      httpClient.onload = () => {
+        // Unsubscribe from the cancellation token.
+        cancellationUnsubscribe();
+
+        let result: TResult | null = null;
+
+        // Check the HTTP response status code to be in the success range.
+        if (httpClient.status >= 200 && httpClient.status <= 299) {
+          result = JSON.parse(httpClient.response);
```
Comment (joluj) [inline frontend/src/services/HTTPHelper.ts:200]: Will fail when header `'Content-Type': 'application/json'` is overwritten.

File: frontend/src/services/HTTPHelper.ts [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+  ): Promise<TResult>;
+
+  public get<TResult>(
+    url: string | URL,
+    cancellation?: CancellationToken
+  ): Promise<TResult>;
+
+  public async get<TResult>(
+    url: string | URL,
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
```
Comment (joluj) [inline frontend/src/services/HTTPHelper.ts:252]: this `eslint-disable-next-line` can be removed.

File: frontend/src/services/HTTPHelper.ts [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+  public post<TResult>(
+    url: string | URL,
+    request: HTTPPOSTRequest,
+    cancellation?: CancellationToken
+  ): Promise<TResult>;
+
+  public post<TResult>(
+    url: string | URL,
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
+    body: any,
```
Comment (joluj) [inline frontend/src/services/HTTPHelper.ts:292]: use `unknown` instead of `any`

File: frontend/src/services/HTTPHelper.ts [hunk 11]
```diff
@@ -0,0 +1,368 @@
...
+    cancellation?: CancellationToken
+  ): Promise<TResult>;
+
+  public async post<TResult>(
+    url: string | URL,
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
+    requestOrBody: HTTPPOSTRequest | any,
```
Comment (joluj) [inline frontend/src/services/HTTPHelper.ts:299]: use `unknown` instead of `any`

File: frontend/src/services/HTTPHelper.ts [hunk 12]
```diff
@@ -0,0 +1,368 @@
...
+      httpResponse.status <= 299 &&
+      httpResponse.result
+    ) {
+      return httpResponse.result;
+    }
+
+    throw Error(httpResponse.statusText);
+  }
+
+  private appendQuery(url: URL | string, query: URLQuery): URL {
```
Comment (joluj) [inline frontend/src/services/HTTPHelper.ts:327]: I'd move this method out of that file and make it only return a query string.
The URL creation with `this.options.baseUrl` is a side effect nobody expects from that function name.
Example:
```ts
createQueryParams(query: URLQuery): string {
     // ...
}

createQueryParams({x: 1, y: ['a', 'b']}) // returns x=1&y=a&y=b
```

File: frontend/src/services/HTTPHelperOptions.ts [hunk 13]
```diff
@@ -0,0 +1,10 @@
+/**
+ * Contains options for the HTTP helper
+ */
+export interface HTTPHelperOptions {
+    /**
+     * The base URI to use.
+     * If not specified, uses the base URI of the source the frontend was loaded from.
+     */
+    baseUri: string;
+}
```
Comment (joluj) [inline frontend/src/services/HTTPHelperOptions.ts:10]: no new line

File: frontend/src/services/HTTPHelper.ts [hunk 14]
```diff
@@ -0,0 +1,368 @@
...
+  // eslint-disable-next-line no-useless-constructor
+  public constructor(
+    public readonly result: TResult | null,
+    public readonly headers: HTTPHeaderCollection,
+    public readonly status: number,
+    public readonly statusText: string
+  ) {}
+}
+
+function parseResponseHeaders(headers: string): HTTPHeaderCollection {
```
Comment (joluj) [inline frontend/src/services/HTTPHelper.ts:62]: Maybe extract this function.
It provides functionality that is not limited to the `HttpHelper` that _may_ (probably not:D) be reused.

File: frontend/src/services/HTTPHelper.ts [hunk 15]
```diff
@@ -0,0 +1,368 @@
...
+    baseUri = `${location.protocol}//${location.host}/${
+      location.pathname.split('/')[1]
+    }`;
+  }
+
+  return { baseUri };
+}
+
+@injectable()
+export default class HTTPHelper {
```
Comment (joluj) [inline frontend/src/services/HTTPHelper.ts:111]: I'd rename it to `HttpService`.

File: frontend/src/services/QueryServiceImpl.ts [hunk 16]
```diff
@@ -1,94 +1,27 @@
...
-  private readonly options: QueryServiceOptions;
-
-  public constructor(options: Partial<QueryServiceOptions> = {}) {
-    super();
-
-    this.options = buildOptions(options);
-  }
+  @inject(HTTPHelper)
+  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+  private readonly http: HTTPHelper = null!;
```
Comment (joluj) [inline frontend/src/services/QueryServiceImpl.ts:17]:
```suggestion
  private readonly http: HTTPHelper | null = null;
```

**Review by joluj — CHANGES_REQUESTED** [review]
Just minor improvements required (don't use `any`-type;).
And I'd thin out `HttpHelper.ts`, since the functions that are declared in this file can be reused and are not only useful in the context of this `HttpHelper`.

**Review by PianoRollRepresentation — APPROVED** [review]

File: frontend/src/services/HTTPHelper.ts (hunk 7 above)
Comment (CatoLeanTruetschel) [reply frontend/src/services/HTTPHelper.ts:172]: That's because we don't have a value for any key. The type would not be correct, as typescript assumes that we can now pass an arbitrary string as key and get a string back. Not for all keys are there values however.

File: frontend/src/services/HTTPHelper.ts (hunk 8 above)
Comment (CatoLeanTruetschel) [reply frontend/src/services/HTTPHelper.ts:200]: That's true, but what's the alternative? A generic value formatter that can be injection? That would be overkill. If the caller wants to shoot themselves in the foot... go for it

File: frontend/src/services/HTTPHelper.ts (hunk 14 above)
Comment (CatoLeanTruetschel) [reply frontend/src/services/HTTPHelper.ts:62]: I don't see value in this currently. If this is ever needed in another place, it is very easy to move it to a new file and export it. Currently this would bloat our directory only.

File: frontend/src/services/QueryServiceImpl.ts (hunk 16 above)
Comment (CatoLeanTruetschel) [reply frontend/src/services/QueryServiceImpl.ts:17]: It is actually guaranteed, that this is non-null, as this gets injected by DI. Sadly I couldn't get constructor injection to work, but property injection does its job here.

File: frontend/src/services/HTTPHelper.ts (hunk 12 above)
Comment (CatoLeanTruetschel) [reply frontend/src/services/HTTPHelper.ts]: This function does also append this to a specified URL, which is non-trivial. But you are right, that the creation of the URL based on the baseUri is not the responsibility of this method.

File: frontend/src/services/HTTPHelper.ts (hunk 14 above)
Comment (joluj) [reply frontend/src/services/HTTPHelper.ts]: It's just a preference that comes from my experience in other languages. I don't like bloated files, I prefer files with only one class oder function over files that are named HttpHelper.ts but where i have to search for the class. My suggestion would bloat the repository, but i think you can set it up so that only the class HttpHelper is visible outside a /HttpHelper folder.

TLDR: Do it like you want :D

File: frontend/src/services/HttpService.ts [hunk 17]
```diff
@@ -0,0 +1,352 @@
+/* eslint-disable max-classes-per-file */
```
Comment (joluj) [inline frontend/src/services/HttpService.ts:1]: This rule is there for a reason. At least move the exported classes to separate files.

Put everything in a single file under a folder `HttpService`.
Create `index.ts` with the content like
```ts
export { HttpService } from './HttpService';
export { HttpRequest } from './HttpRequest';
...
```

and use the imports like that:
```ts
import { HttpService, HttpRequest } from './services/HttpService'
```

Looks the same from the outside, but much cleaner if you look into the files.

**Review by joluj — CHANGES_REQUESTED** [review]
Don't just disable every tslint config.
I understand what you try to achieve with the multi-class-export in `HttpService.ts`, but I think the approach with separate files and an `index.ts` is smoother.

**Review by joluj — APPROVED** [review]

**Review by joluj — APPROVED** [review]

**Comment by derwehr** [comment]
I believe this merge broke `http://localhost:3000/visualization/graph`

On this branch it outputs:
```
Something went wrong: The response contains a failure error code: 404 Not Found
```

**Comment by CatoLeanTruetschel** [comment]
On my machine everything works fine. Is the backend running when you perform the query?

**Comment by joluj** [comment]
@CatoLeanTruetschel can you change  `.env.example`, since the variable was renamed to`REACT_APP_QUERY_SERVICE_BASE_URI`?

**Comment by derwehr** [comment]
> On my machine everything works fine. Is the backend running when you perform the query?
Changing the variable name in my run script did the trick, thank you.

**Comment by CatoLeanTruetschel** [comment]
> can you change .env.example, since the variable was renamed toREACT_APP_QUERY_SERVICE_BASE_URI?
Sure, this is tracked by #122
